decentralize-authz-app/src/repositories/kv-repository.ts

```diff
@@ -0,0 +1,186 @@
+      const value: any = this.kvStore.get(ccf.strToBuf(key));
+     
+      if (value === undefined) {
+        return ServiceResult.Failed({
+          errorMessage: "Error: key does not exist",
+          errorType: "KeyValueStoreError",
+        });
+      }
+
+      const data = ccf.bufToJsonCompatible(value) as T;
+
+      return ServiceResult.Succeeded(data);
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to read value from the kvstore",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public has(key: string): ServiceResult<boolean> {
+    try {
+      return ServiceResult.Succeeded(this.kvStore.has(ccf.strToBuf(key)));
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to check if key exists in the kvstore",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public keys(): ServiceResult<string[]> {
+    try {
+      const keys: string[] = [];
+      this.kvStore.forEach((val, key) => {
+        keys.push(ccfapp.bufToStr(key));
+      });
+      return ServiceResult.Succeeded(keys);
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to get kvstore all keys",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public values(): ServiceResult<T[]> {
+    try {
+      const values: T[] = [];
+      this.kvStore.forEach((val, key) => {
+        values.push(val);
+      });
+      return ServiceResult.Succeeded(values);
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to get kvstore all values",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public clear(): ServiceResult<void> {
+    try {
+      return ServiceResult.Succeeded(this.kvStore.clear());
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to clear kvstore values",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public delete(key: string): ServiceResult<void> {
+    try {
+      return ServiceResult.Succeeded(this.kvStore.delete(ccf.strToBuf(key)));
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to remove a key",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  // 
+  public forEach(callback: (key: string, value: T) => void): ServiceResult<string> {
+    try {
+      this.kvStore.forEach((val, key) => {
+        callback(ccf.bufToStr(key), val);
+      });
+
+      return ServiceResult.Succeeded("");
+    } catch (ex) {
+      return ServiceResult.Failed({
+        errorMessage: "Error: unable to iterate the kvstore pairs",
+        errorType: "KeyValueStoreError",
+        details: ex,
+      });
+    }
+  }
+
+  public get size(): ServiceResult<number> {
```

Comment:
Unless I am missing something, these wrappers largely duplicate functionality that's implemented in CCF, where exceptions are caught at the top level, and reported in a 500 to the user (unless caught directly and handled by the business logic). There are centralised controls over whether the exception details are logged and/or returned via a governance controlled table (https://microsoft.github.io/CCF/main/audit/builtin_maps.html#_CPPv4N3ccf16JSRuntimeOptionsE), whereas this seems to always return full details.
Are we sure we need them?

Reply:
I removed the part where the exception is returned to the user and logged it locally.